sdk/core/System.ClientModel/src/Internal/ArrayBackedPropertyBag.cs

```diff
@@ -0,0 +1,308 @@
+    {
+        switch (_count)
+        {
+            case 0:
+                _first = new Kvp(key, value);
+                _count = 1;
+                return;
+            case 1:
+                if (IsFirst(key))
+                {
+                    _first = new Kvp(_first.Key, value);
+                }
+                else
+                {
+                    _second = new Kvp(key, value);
+                    _count = 2;
+                }
+
+                return;
+            default:
+                if (_rest == null)
+                {
+                    _rest = ArrayPool<Kvp>.Shared.Rent(8);
+                    _rest[_count++ - 2] = new Kvp(key, value);
+                    return;
+                }
+
+                if (_rest.Length <= _count)
+                {
+                    var larger = ArrayPool<Kvp>.Shared.Rent(_rest.Length << 1);
+                    _rest.CopyTo(larger, 0);
+                    var old = _rest;
+                    _rest = larger;
+                    ArrayPool<Kvp>.Shared.Return(old, true);
+                }
+                _rest[_count++ - 2] = new Kvp(key, value);
+                return;
+        }
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private void SetAt(int index, Kvp value)
+    {
+        if (index == 0)
+            _first = value;
+        else if (index == 1)
+            _second = value;
+        else
+            GetRest()[index - 2] = value;
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private TValue GetAt(int index) => index switch
+    {
+        0 => _first.Value,
+        1 => _second.Value,
+        _ => GetRest()[index - 2].Value
+    };
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private int GetIndex(TKey key)
+    {
+        if (_count == 0)
+            return -1;
+        if (_count > 0 && _first.Key.Equals(key))
+            return 0;
+        if (_count > 1 && _second.Key.Equals(key))
+            return 1;
+
+        if (_count <= 2)
+            return -1;
+
+        Kvp[] rest = GetRest();
+        int max = _count - 2;
+        for (var i = 0; i < max; i++)
+        {
+            if (rest[i].Key.Equals(key))
+                return i + 2;
+        }
+        return -1;
+    }
+
+    public void Dispose()
+    {
+#if DEBUG
+        if (_disposed)
+        {
+            return;
+        }
+        _disposed = true;
+#endif
+        _count = 0;
+        _first = default;
+        _second = default;
+        if (_rest == default)
+        {
+            return;
+        }
+
+        var rest = _rest;
+        _rest = default;
+        ArrayPool<Kvp>.Shared.Return(rest, true);
+    }
+
+    private Kvp[] GetRest() => _rest ?? throw new InvalidOperationException($"{nameof(_rest)} field is null while {nameof(_count)} == {_count}");
+
+#pragma warning disable CA1822
+    [Conditional("DEBUG")]
+    private void CheckDisposed()
+    {
+#if DEBUG
```

Comment:
is this needed given the method itself is conditional?

Reply:
That is a good point.  Since we use the compilation variables elsewhere, I'll remove the attribute.